In [1]:
from platform import python_version
print('Versão do pyhton utilizada neste notebook: {} '.format(python_version()))

Versão do pyhton utilizada neste notebook: 3.12.4 


# Importações Necessários

In [25]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Lendo o arquivo com os resultados anteriores

In [3]:
df = pd.read_csv('lotofacil.csv')

# Trazendo as primeira linhas do DataFrame

In [4]:
df.head()

Concurso  Bola1  Bola2  Bola3  Bola4  Bola5  Bola6  Bola7  Bola8  Bola9  \
0      3179      3      4      5      7      8     11     12     16     17   
1      3178      1      3      4      5      6      7      9     10     14   
2      3177      1      3      5      6      7      8     10     14     15   
3      3176      1      2      3      6      7      8     11     12     14   
4      3175      1      2      3      5      6      7      8      9     10   

   Bola10  Bola11  Bola12  Bola13  Bola14  Bola15  \
0      18      21      22      23      24      25   
1      16      19      20      23      24      25   
2      18      19      20      21      23      24   
3      15      17      21      22      24      25   
4      11      13      17      22      23      24   

                                 Resultado  
0  3,4,5,7,8,11,12,16,17,18,21,22,23,24,25  
1    1,3,4,5,6,7,9,10,14,16,19,20,23,24,25  
2   1,3,5,6,7,8,10,14,15,18,19,20,21,23,24  
3   1,2,3,6,7,8,11,12,14,15,17,21,22,24,25  
4     1,2,3,5,6,7,8,9,10,11,13,17,22,23,24

# Convertendo a coluna "Resultado" em uma lista de números

In [6]:
df['Resultado'] = df['Resultado'].apply(lambda x : list(map(int,x.split(','))))

In [7]:
df.head()

Concurso  Bola1  Bola2  Bola3  Bola4  Bola5  Bola6  Bola7  Bola8  Bola9  \
0      3179      3      4      5      7      8     11     12     16     17   
1      3178      1      3      4      5      6      7      9     10     14   
2      3177      1      3      5      6      7      8     10     14     15   
3      3176      1      2      3      6      7      8     11     12     14   
4      3175      1      2      3      5      6      7      8      9     10   

   Bola10  Bola11  Bola12  Bola13  Bola14  Bola15  \
0      18      21      22      23      24      25   
1      16      19      20      23      24      25   
2      18      19      20      21      23      24   
3      15      17      21      22      24      25   
4      11      13      17      22      23      24   

                                           Resultado  
0  [3, 4, 5, 7, 8, 11, 12, 16, 17, 18, 21, 22, 23...  
1  [1, 3, 4, 5, 6, 7, 9, 10, 14, 16, 19, 20, 23, ...  
2  [1, 3, 5, 6, 7, 8, 10, 14, 15, 18, 19, 20, 21,...  
3  [1, 2, 3, 6, 7, 8, 11, 12, 14, 15, 17, 21, 22,...  
4  [1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 13, 17, 22, 2...

# Definindo as carateristicas e rótulos que serão analisados

In [8]:
X = []
y = []

In [9]:
for i in range(1,len(df)):
    X.append(df['Resultado'][i-1])
    y.append(df['Resultado'][i])

#Convertendo para DataFrame
X = pd.DataFrame(X)
y = pd.DataFrame(y)

# Dividir os dados em conjuntos de treinamento e teste

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Mostrar o tamanho dos conjuntos de treinamento e teste

In [37]:
print(f'Tamanho do conjunto de treinamento: {X_train.shape[0]}')
print(f'Tamanho do conjunto de teste: {X_test.shape[0]}')

Tamanho do conjunto de treinamento: 2534
Tamanho do conjunto de teste: 634


# Modelo de Previsão

In [10]:
#Iniciando o modelo
modelo = LinearRegression()

In [11]:
#Treinando o modelo
modelo.fit(X_train, y_train)

LinearRegression()

# Validando o modelo - Erro Quadrático Médio (MSE)

    y_test: Representa os valores reais dos rótulos (labels) no conjunto de teste. Esses são os valores que o modelo deve prever corretamente.
    
y_pred: Representa os valores previstos pelo modelo para o conjunto de teste. Esses são os valores que o modelo gerou com base nos dados de     treinamento
    .

In [19]:
mse = mean_squared_error(y_test, y_pred)
print(f'Erro Quadrático Médio (MSE): {mse}')

Erro Quadrático Médio (MSE): 2.6098584296575646


# Comparando com um Modelo de Baseline

In [26]:
# Calcular a média dos valores de treinamento
mean_y_train = np.mean(y_train)

In [27]:
# Prever a média para todos os valores de teste
y_pred_baseline = np.full_like(y_test, mean_y_train)

In [28]:
# Calcular o MSE do modelo baseline
mse_baseline = mean_squared_error(y_test, y_pred_baseline)
print(f'Erro Quadrático Médio (MSE) do Modelo Baseline: {mse_baseline}')

Erro Quadrático Médio (MSE) do Modelo Baseline: 52.62891692954784


# Resumo da comparação
O Erro Quadrático Médio (MSE) do modelo baseline é 52.62891692954784, enquanto o MSE do seu modelo de regressão linear é 2.6098584296575646. Isso indica que o seu modelo de regressão linear está performando significativamente melhor do que o modelo baseline, o que é um bom sinal.
**
Interpretação dos Resultad**
o    1.s
Comparação com o Modelo Baseline: O MSE do seu modelo (2.61) é muito menor do que o MSE do modelo baseline (52.63). Isso sugere que o seu modelo está fazendo previsões mais precisas em comparação com um modelo que simplesmente prevê a média dos valores de treinamento    2..

Escala dos Dados: Se os valores dos rótulos (labels) estão na faixa de 1 a 25 (como é comum em sorteios de loteria), um MSE de aproximadamente 2.61 é relativamente baixo, indicando que as previsões do modelo estão próximas dos valores rea    3.is.

Contexto do Problema: Embora o MSE do seu modelo seja baixo em comparação com o modelo baseline, é importante lembrar que a previsão de números de loteria é altamente incerta devido à natureza aleatória dos sorteios. Portanto, mesmo um MSE baixo não garante previsões precisas.

# Prevendo o próximo resultado

In [22]:
ultimo_resultado = df['Resultado'].iloc[-1]
previsao = modelo.predict([ultimo_resultado])

# Converter a previsão para uma lista de inteiros
previsao = previsao[0].astype(int).tolist()

# Mostrando a previsão

In [23]:
print(f'Previsão para o concurso 3180: {previsao}')

Previsão para o concurso 3180: [1, 3, 4, 6, 8, 9, 11, 12, 14, 16, 17, 19, 21, 22, 24]


# Adicionar a nova linha ao DataFrame existente

In [29]:
nova_linha = pd.DataFrame({
    'Concurso': [3180],
    'Bola1': [previsao[0]],
    'Bola2': [previsao[1]],
    'Bola3': [previsao[2]],
    'Bola4': [previsao[3]],
    'Bola5': [previsao[4]],
    'Bola6': [previsao[5]],
    'Bola7': [previsao[6]],
    'Bola8': [previsao[7]],
    'Bola9': [previsao[8]],
    'Bola10': [previsao[9]],
    'Bola11': [previsao[10]],
    'Bola12': [previsao[11]],
    'Bola13': [previsao[12]],
    'Bola14': [previsao[13]],
    'Bola15': [previsao[14]],
    'Resultado': [','.join(map(str, previsao))]
})

In [34]:
df = pd.concat([df, pd.DataFrame(nova_linha)], ignore_index=True)

In [35]:
df

Concurso  Bola1  Bola2  Bola3  Bola4  Bola5  Bola6  Bola7  Bola8  Bola9  \
0         3179      3      4      5      7      8     11     12     16     17   
1         3178      1      3      4      5      6      7      9     10     14   
2         3177      1      3      5      6      7      8     10     14     15   
3         3176      1      2      3      6      7      8     11     12     14   
4         3175      1      2      3      5      6      7      8      9     10   
...        ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3165         4      1      2      4      5      8     10     12     13     16   
3166         3      1      4      6      7      8      9     10     11     12   
3167         2      1      4      5      6      7      9     11     12     13   
3168         1      2      3      5      6      9     10     11     13     14   
3169      3180      1      3      4      6      8      9     11     12     14   

      Bola10  Bola11  Bola12  Bola13  Bola14  Bola15  \
0         18      21      22      23      24      25   
1         16      19      20      23      24      25   
2         18      19      20      21      23      24   
3         15      17      21      22      24      25   
4         11      13      17      22      23      24   
...      ...     ...     ...     ...     ...     ...   
3165      17      18      19      23      24      25   
3166      14      16      17      20      23      24   
3167      15      16      19      20      23      24   
3168      16      18      20      23      24      25   
3169      16      17      19      21      22      24   

                                              Resultado  
0     [3, 4, 5, 7, 8, 11, 12, 16, 17, 18, 21, 22, 23...  
1     [1, 3, 4, 5, 6, 7, 9, 10, 14, 16, 19, 20, 23, ...  
2     [1, 3, 5, 6, 7, 8, 10, 14, 15, 18, 19, 20, 21,...  
3     [1, 2, 3, 6, 7, 8, 11, 12, 14, 15, 17, 21, 22,...  
4     [1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 13, 17, 22, 2...  
...                                                 ...  
3165  [1, 2, 4, 5, 8, 10, 12, 13, 16, 17, 18, 19, 23...  
3166  [1, 4, 6, 7, 8, 9, 10, 11, 12, 14, 16, 17, 20,...  
3167  [1, 4, 5, 6, 7, 9, 11, 12, 13, 15, 16, 19, 20,...  
3168  [2, 3, 5, 6, 9, 10, 11, 13, 14, 16, 18, 20, 23...  
3169             1,3,4,6,8,9,11,12,14,16,17,19,21,22,24  

[3170 rows x 17 columns]

# Salvando o DataFrame com o resultado 3180

In [36]:
df.to_csv('concurso3180.csv', index= False)